target/linux/ar71xx/image/tp-link.mk

```diff
@@ -880,6 +880,21 @@ define Device/tl-wr1043nd-v4
 endef
 TARGET_DEVICES += tl-wr1043nd-v1 tl-wr1043nd-v2 tl-wr1043nd-v3 tl-wr1043nd-v4
 
+define Device/tl-wr942n-v1
+  DEVICE_TITLE := TP-LINK TL-WR942N v1
+  DEVICE_PACKAGES := kmod-usb-core kmod-usb2 kmod-usb-ledtrig-usbport
+  BOARDNAME := TL-WR942N-V1
+  TPLINK_BOARD_NAME := TL-WR942N
+  DEVICE_PROFILE := TLWR942NV1
+  IMAGE_SIZE := 14464k
+  KERNEL := kernel-bin | patch-cmdline | lzma | uImageArcher lzma
+  IMAGES := sysupgrade.bin factory.bin
+  IMAGE/sysupgrade.bin := append-rootfs | tplink-safeloader sysupgrade
+  IMAGE/factory.bin := append-rootfs | tplink-safeloader factory
+  MTDPARTS := spi0.0:128k(u-boot)ro,1344k(kernel),13120k(rootfs),64k(product-info)ro,64k(partition-table)ro,64k(soft-version)ro,64k(support-list)ro,640k(config)ro,576k(certificate)ro,64k(usb-config)ro,128k(log)ro,64k(radio-bk)ro,64k(ART)ro,14464k@0x20000(firmware)
```

Comment:
> I would mostly like to be as close as possible to new tp-link "pharos" so called partitioning schema. It is also used during their new multypart image firmware update process and is fully defined in this file for other tplink devices like tl-wr1043 v4, eap120, re450, cpe510/520. Why whould you recomend to use simplified one which is used on C59 only?

First of all, all of that additional partitions are marked as **read-only** which means that even if you want, you can't change their content during the upgrade. If it's possible (take this as a general rule), we never touch/change/overwrite vendor additional data, that's the reason why we keep it in read-only marked partitions.

Secondly, if you look into the [tools/firmware-utils/src/tplink-safeloader.c](https://github.com/lede-project/source/blob/master/tools/firmware-utils/src/tplink-safeloader.c#L134), for CPE510 we don't define all partitions in kernel command line, they are combined in `product-info` and `config` partitions.

> IMHO This simplification will make reverting to factory images much harder if we adopt full support for multypart fw image as some this partitions needs to be owerwritten.

Please refer to tplink-safeloader.c, we don't touch other partitions than kernel and rootfs (os-image and file-system) during the upgrade.

> As for now reverting by LEDE sysupgrade to oem firmware is possible only to the same fw which was before LEDE if previously dumped oem partition "firmware" is used as sisupgrade image. Attempt to revert with different oem version dump leeds to non fully functional oem WEB interface. So this problem has to be learned and discused.

Are there any differences in partition layout between different vendor firmware images?